src/test/java/com/kodilla/ecommercee/domain/CartEntityTestSuite.java

```diff
@@ -0,0 +1,138 @@
+package com.kodilla.ecommercee.domain;
+
+import com.kodilla.ecommercee.exception.CartNotFoundException;
+import com.kodilla.ecommercee.repository.CartRepository;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import javax.transaction.Transactional;
+import java.math.BigDecimal;
+import java.util.List;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class CartEntityTestSuite {
+
+    @Autowired
+    CartRepository cr;
+
+    //Create
+    @Test
+    @Transactional
+    public void testCreateCart() {
+
+        //Given
+        Product p1 = new Product("p1", "Description p1", new BigDecimal(55.55));
+        Product p2 = new Product("p2", "Description p2", new BigDecimal(55.55));
+        Product p3 = new Product("p3", "Description p3", new BigDecimal(55.55));
+        Cart testCart = new Cart();
+
+        //When
+        testCart.getProducts().add(p1);
+        testCart.getProducts().add(p2);
+        testCart.getProducts().add(p3);
+        cr.save(testCart);
+
+        int quantity = cr.findAll().size();
+        int expectedResult = 1;
+        long productQuantityInDatabase = cr.findAll().stream()
+                .flatMap(list -> list.getProducts().stream())
+                .count();
+        long expectedResult2 = 3;
+
+        //Then
+        Assert.assertEquals(expectedResult, quantity);
+        Assert.assertEquals(expectedResult2, productQuantityInDatabase);
+    }
+
+    @Test
+    @Transactional
+    public void testGetCart() {
+
+        //Given
+        Product p1 = new Product("p1", "Description p1", new BigDecimal(55.55));
+        Product p2 = new Product("p2", "Description p2", new BigDecimal(55.55));
+        Product p3 = new Product("p3", "Description p3", new BigDecimal(55.55));
+        Cart testCart = new Cart();
+
+        //When
+        testCart.getProducts().add(p1);
+        testCart.getProducts().add(p2);
+        testCart.getProducts().add(p3);
+
+        cr.save(testCart);
+        List<Cart> cartList = cr.findAll();
+
+//        Double cartValue = cr.findById(1L).orElseThrow(CartNotFoundException::new)
+//                .getProducts()
+//                .stream()
+//                .map(product -> product.getPrice())
+//                .mapToDouble(BigDecimal::doubleValue)
+//                .sum();
+//
+//        Double expected = 166.6;
+
+        //Then
+
+ //       Assert.assertEquals(expected, cartValue, 0.1);
```

Comment:
Ten test po odkomentowaniu u mnie się wykonuje, jednak projekt nie chce wstać, zastanawia mnie o co kaman?

Reply:
u mnie ten test po odkomentowaniu się nie wykonuje ponieważ cr nie może znaleźć koszyka o podanym przez ciebie id. w tym przypadku nie wiesz jakie id zostało wygenerowane dla `testCart`, więc zamiast 1L powinienaś pobrać id które zostało przypisane do `testCart`.

Reply:
Mamy autonumeracje w Encji, i H2 zaczyna nie od 0 tylko od 1. Tabela wyczyszczona więc musi to być 1L (na tym polega test że coś zakładamy) i generalnie to powinno funkcjonować z założeń o których wiem, pytanie czego nie wiem lub o czym nie pomyslałem...

Reply:
Kwestia numeracji. Zrobiłem "bezczelne" System.out.println("Cart id: " + testCart.getId()); którego wynikiem było:  "Cart id: 6"

Reply:
No i właśnie ja też na 3 różnych maszynach i ciągle wychodzi mi wspomniana jedynka. No nic zgodnie z zaleceniem jutro wrzucę poprawione. Tak jak mówiłem bardziej ludzka ciekawość i gdzie leży błąd z czego wynika rozbieżność... Wytyczne jasne ma działać i się kompilować ;) dzięki chłopaki za review

Reply:
Głowy nie dam, ale wydaje mi się, że może chodzić o tabelę hibernate_sequence. Kiedy robiłem testy w innym projekcie, to mimo iż obiekty były usuwane z bazy, to ich id były wciąż "zajęte" i nowe obiekty dostawały kolejne id. Dopiero drop tabeli hibernate_sequence powodował "wyzerowanie" i przydzielanie id od "1". Ale było to w przypadku bazy SQL, nie H2.

Reply:
w tej kwestii to ja bym nie był taki krytyczny względem @przemyslawkrawczynski . Natomiast @przemyslawkrawczynski  też powinieneś wziąć pod uwagę, że u kogoś może coś nie przechodzić. Tutaj to jest klasyczna standardowa odpowiedź administratora numer 1 (SOA1), czyli "u mnie działa". Generalnie fajnie by było gdyby testy wykonywały się zawsze - niezależnie od stanu bazy. Więc albo sprawdzamy ID, który został zwrócony z bazy danych albo robimy pełny tearUp() i tearDown(). Ogólnie unit testy robicie na H2 ale może będziecie chcieli kiedyś odpalić pełną integrację na rzeczywistej bazie jak Oracle czy MySQL? Implementacje testów macie już gotową i gdyby testy nie były zależne od stanu bazy to wtedy możemy podmienić tylko "sterownik" na MySQL i zrobić pełne testy integracyjne z produkcyjną bazą danych... wtedy możemy np. dodać do pipeline w CI/CD albo ustawić, że ma lecieć raz dziennie w ciągu nocy na środowisku przed-produkcyjnym.  

Mam nadzieję, że trochę światła rzuciłem w tym temacie :-)